### PR TITLE
 Add operator cache and coercion cache to reduce planning time

### PR DESCRIPTION
**Pull request by lzeiming**
MetadataManager.resolveOperator() and MetadataManager.getCoercion() are frequently invoked during planning.
Add cache for them could reduce many planning time for IN predicate with a large list.

### before the change
Benchmark                            (stage)  Mode  Cnt      Score     Error  Units
BenchmarkPlanner.planLargeInQuery  optimized  avgt   20  17808.872 ± 326.616  ms/op
BenchmarkPlanner.planLargeInQuery    created  avgt   20     52.415 ±   2.171  ms/op

### After the change
Benchmark                            (stage)  Mode  Cnt     Score    Error  Units
BenchmarkPlanner.planLargeInQuery  optimized  avgt   20  5110.045 ± 88.355  ms/op
BenchmarkPlanner.planLargeInQuery    created  avgt   20    50.761 ±  1.200  ms/op

### CODE REVIEW

File: core/trino-main/src/test/java/io/trino/sql/planner/BenchmarkPlanner.java [hunk 1]
```diff
@@ -125,6 +134,16 @@ public String readResource(String resource)
         });
     }
 
+    @Benchmark
+    public Plan planLargeInQuery(BenchmarkData benchmarkData)
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/sql/planner/BenchmarkPlanner.java:138]: add a test case to `verify`

File: core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java [hunk 2]
```diff
@@ -201,6 +209,9 @@
 
     private final ResolvedFunctionDecoder functionDecoder;
 
+    private final LoadingCache<Entry<OperatorType, List<? extends Type>>, ResolvedFunction> operatorCache;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:212]: please create a dedicated cache key classes.

File: core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java [hunk 3]
```diff
@@ -248,6 +259,23 @@ public MetadataManager(
...
+                }));
+
+        coercionCache = CacheBuilder.newBuilder()
+                .maximumSize(1000)
+                .build(CacheLoader.from(cell -> {
+                    String name = mangleOperatorName(cell.getRowKey());
+                    Type fromType = cell.getColumnKey();
+                    Type toType = cell.getValue();
+                    Signature signature = new Signature(name, toType.getTypeSignature(), ImmutableList.of(fromType.getTypeSignature()));
+                    return resolve(functionResolver.resolveCoercion(functions.get(QualifiedName.of(name)), signature));
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:277]: You can just cache `io.trino.metadata.MetadataManager#resolve(io.trino.metadata.FunctionBinding, io.trino.metadata.FunctionDependencyDeclaration)` as all other functions delegate to it eventually.

File: core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java [hunk 4]
```diff
@@ -15,11 +15,18 @@
 
 import com.google.common.annotations.VisibleForTesting;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:16]: 1. Please add explanation in commit message why do you cache resolved function and what was the biggest performance issue.
2. Please make benchmarks results part of commit message

File: core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java [hunk 5]
```diff
@@ -248,6 +259,23 @@ public MetadataManager(
...
+
+        operatorCache = CacheBuilder.newBuilder()
+                .maximumSize(1000)
+                .build(CacheLoader.from(entry -> {
+                    String name = mangleOperatorName(entry.getKey());
+                    return resolveFunction(QualifiedName.of(name), fromTypes(entry.getValue()));
+                }));
+
+        coercionCache = CacheBuilder.newBuilder()
+                .maximumSize(1000)
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:271]: Do we want (or need?) to make this cache part of transaction state?

Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:271]: Also, question about sizing. I don't know what would be the proper size here.

eg how many different entries are getting inserted here during planning of the TPC-H and TPC-DS queries?

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:271]: I think it's transaction agnostic

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:271]: We have similar cache in `io.trino.metadata.FunctionRegistry#FunctionRegistry`

**Comment by martint** [comment]
I think @dain has some pending changes related to this in the work he’s been doing around function system revamp

File: core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java (hunk 3 above)
Comment (lzeiming) [reply core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:277]: Then FunctionResolver.resolveCoercion() is not cached. 
From what I observed, FunctionResolver.resolveFunction() and FunctionResolver.resolveCoercion() are the slow part. 
I considered adding cache in FunctionResolver, but some of the argument types are not suitable to be a part of cache key, such as TypeSignatureProvider.

File: core/trino-main/src/test/java/io/trino/sql/planner/BenchmarkPlanner.java (hunk 1 above)
Comment (lzeiming) [reply core/trino-main/src/test/java/io/trino/sql/planner/BenchmarkPlanner.java:139]: Add in verify()

File: core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java [hunk 6]
```diff
@@ -2314,4 +2348,97 @@ private synchronized void finish()
...
+
+    private static class OperatorCacheKey
+    {
+        private final OperatorType operatorType;
+        private final List<? extends Type> argumentTypes;
+
+        private OperatorCacheKey(OperatorType operatorType, List<? extends Type> argumentTypes)
+        {
+            this.operatorType = requireNonNull(operatorType, "operatorType is null");
+            requireNonNull(argumentTypes, "argumentTypes is null");
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:2360]: inline inside `copyOf`

File: core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java [hunk 7]
```diff
@@ -1884,12 +1909,17 @@ public ResolvedFunction resolveOperator(OperatorType operatorType, List<? extend
...
-                throw new OperatorNotFoundException(operatorType, argumentTypes, e);
+        catch (UncheckedExecutionException e) {
+            if (e.getCause() instanceof TrinoException) {
+                TrinoException cause = (TrinoException) e.getCause();
+                if (cause.getErrorCode().getCode() == FUNCTION_NOT_FOUND.toErrorCode().getCode()) {
+                    throw new OperatorNotFoundException(operatorType, argumentTypes, cause);
+                }
+                throw cause;
             }
+            throwIfInstanceOf(e.getCause(), RuntimeException.class);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:1922]: we can skip this

File: core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java [hunk 8]
```diff
@@ -1899,13 +1929,17 @@ public ResolvedFunction getCoercion(OperatorType operatorType, Type fromType, Ty
...
-                throw new OperatorNotFoundException(operatorType, ImmutableList.of(fromType), toType.getTypeSignature(), e);
+        catch (UncheckedExecutionException e) {
+            if (e.getCause() instanceof TrinoException) {
+                TrinoException cause = (TrinoException) e.getCause();
+                if (cause.getErrorCode().getCode() == FUNCTION_IMPLEMENTATION_MISSING.toErrorCode().getCode()) {
+                    throw new OperatorNotFoundException(operatorType, ImmutableList.of(fromType), toType.getTypeSignature(), cause);
+                }
+                throw cause;
             }
+            throwIfInstanceOf(e.getCause(), RuntimeException.class);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:1942]: we can skip this

**Review by sopel39 — APPROVED** [review]
lgtm % small comments

**Comment by sopel39** [comment]
merged, thanks!

**Pull request merged**